int32 num_moves
---
int32 moves
---
int32 current_move